test/tools/cmap_spec_runner.ts

```diff
@@ -0,0 +1,336 @@
+import { expect } from 'chai';
+import { EventEmitter } from 'events';
+import * as util from 'util';
+
+import { Connection } from '../../src';
+import { ConnectionPool } from '../../src/cmap/connection_pool';
+import { sleep } from './utils';
+
+type cmapOperation =
+  | { name: 'start' | 'waitForThread'; target: string }
+  | { name: 'wait'; ms: number }
+  | { name: 'waitForEvent'; event: string; count: number }
+  | { name: 'checkOut'; thread: string; label: string }
+  | { name: 'checkIn'; connection: string }
+  | { name: 'clear' | 'close' | 'ready' };
+
+type cmapPoolOptions = {
+  maxPoolSize?: number;
+  minPoolSize?: number;
+  maxIdleTimeMS?: number;
+  waitQueueTimeoutMS?: number;
+};
+
+type cmapEvent = {
+  type: string;
+  address?: 42;
+  connectionId?: number;
+  options?: 42 | cmapPoolOptions;
+  reason: string;
+};
+
+const knownTestKeys = [
+  'name',
+  'version',
+  'style',
+  'description',
+  'poolOptions',
+  'operations',
+  'error',
+  'events',
+  'ignore'
+];
+export type cmapTest = {
+  name?: string; // filename path added by the spec loader
+  version: number;
+  style: 'unit';
+  description: string;
+  poolOptions?: cmapPoolOptions;
+  operations: cmapOperation[];
+  error?: {
+    type: string;
+    message: string;
+    address?: number;
+  };
+  events?: cmapEvent[];
+  ignore?: string[];
+};
+
+const ALL_POOL_EVENTS = new Set([
+  ConnectionPool.CONNECTION_POOL_CREATED,
+  ConnectionPool.CONNECTION_POOL_CLOSED,
+  ConnectionPool.CONNECTION_POOL_CLEARED,
+  ConnectionPool.CONNECTION_CREATED,
+  ConnectionPool.CONNECTION_READY,
+  ConnectionPool.CONNECTION_CLOSED,
+  ConnectionPool.CONNECTION_CHECK_OUT_STARTED,
+  ConnectionPool.CONNECTION_CHECK_OUT_FAILED,
+  ConnectionPool.CONNECTION_CHECKED_OUT,
+  ConnectionPool.CONNECTION_CHECKED_IN
+]);
+
+function getEventType(event) {
+  const eventName = event.constructor.name;
+  return eventName.substring(0, eventName.lastIndexOf('Event'));
+}
+
+/**
+ * In the cmap spec and runner definition,
+ * a "thread" refers to a concurrent execution context
+ */
+class Thread {
+  _promise: Promise<void>;
+  _error: Error;
+  _killed = false;
+
+  _knownCommands: any;
+
+  // concurrent execution context
+  constructor(operations) {
+    this._promise = new Promise(resolve => {
+      this.start = () => resolve();
+    });
+
+    this._knownCommands = operations;
+  }
+
+  start: () => void;
+
+  queue(op: cmapOperation) {
+    if (this._killed || this._error) {
+      return;
+    }
+
+    this._promise = this._promise.then(() => this._runOperation(op)).catch(e => (this._error = e));
```

Comment:
I left this logic untouched from the original implementation; I can update it to remove the unnecessary assignment, but I don't think the mongo error bit is necessary since we do check the error in the assertions later on

Reply:
@nbbeeken I actually double checked this, and, per the [official docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises):
> Here's the magic: the then() function returns a new promise, different from the original